components/brave_news/browser/feed_v2_builder.cc

```diff
@@ -1122,79 +1125,138 @@ void FeedV2Builder::NotifyUpdateCompleted() {
+                          feed->construct_time = base::Time::Now();
+                          feed->type = std::move(type);
+
+                          if (!builder) {
+                            std::move(callback).Run(std::move(feed));
+                            return;
+                          }
+
+                          feed->source_hash = builder->hash_;
+
+                          if (feed->items.empty()) {
+                            // If we have no subscribed items and we've loaded
+                            // the list of publishers (which we might not have,
+                            // if we're offline) then we're not subscribed to
+                            // any feeds.
+                            if (builder->subscribed_count_ == 0 &&
+                                !builder->publishers_controller_
+                                     ->GetLastPublishers()
+                                     .empty()) {
+                              feed->error = mojom::FeedV2Error::NoFeeds;
+                              // If we don't have any raw feed items (and we're
+                              // subscribed to some feeds) then fetching must
+                              // have failed.
+                            } else if (builder->raw_feed_items_.size() == 0) {
+                              feed->error = mojom::FeedV2Error::ConnectionError;
+                              // Otherwise, this feed must have no articles.
+                            } else {
+                              feed->error = mojom::FeedV2Error::NoArticles;
+                            }
+                          }
+
+                          std::move(callback).Run(std::move(feed));
+                        },
+                        builder, std::move(type), std::move(built_callback)));
           },
-          weak_ptr_factory_.GetWeakPtr(), std::move(type),
-          std::move(build_feed), std::move(callback)));
+          weak_ptr_factory_.GetWeakPtr(), std::move(type), std::move(generate),
+          std::move(callback)));
 }
 
-mojom::FeedV2Ptr FeedV2Builder::GenerateBasicFeed(const FeedItems& feed_items,
+mojom::FeedV2Ptr FeedV2Builder::GenerateBasicFeed(const FeedV2Builder& builder,
+                                                  const FeedItems& feed_items,
                                                   PickArticles pick_hero,
                                                   PickArticles pick_article) {
   DVLOG(1) << __FUNCTION__;
 
-  auto suggested_publisher_ids = suggested_publisher_ids_;
-  auto articles =
-      GetArticleInfos(publishers_controller_->GetLastLocale(), feed_items,
-                      publishers_controller_->GetLastPublishers(), signals_);
+  auto suggested_publisher_ids = builder.suggested_publisher_ids_;
+  auto articles = GetArticleInfos(
+      builder.publishers_controller_->GetLastLocale(), feed_items,
+      builder.publishers_controller_->GetLastPublishers(), builder.signals_);
```

Comment:
Regardless of logical invariant of the builder's state,
to me it feel like accessing builder's member variable and publisher_controller is exposing us to potential risk of thread safety.  If we're to do this, PublisherController's GetLastLocale() and others should be guaranteed of thread safety. 

I think bind all recipes like publisher ids, locale, last publishers, signal from the main thread would be much safer.

(GenerateAllfeed as well)

Reply:
Okay cool, I was considering that but got scared off by the fact that we were doing this for perf. Honestly, it'd make me feel a lot better about this to do it that way :+1:

Reply:
> got scared off by the fact that we were doing this for perf

Yeah, now I see where you're coming from. Your concern makes sense if all vector is huge. I think we can try using RefCountedThreadSafe and others. Let me try a bit.

Reply:
I suggest not use `RefCountedThreadSafe`, it makes the code much harder to maintain in terms of avoiding data races.

There are a few alternatives:
1. Post everything each time you need as extra arguments as a copy in function/static method/lambda (depends on the data, could be costly).
2. Make a task runner inner class with necessary data. See FooService here: https://source.chromium.org/chromium/chromium/src/+/main:docs/threading_and_tasks_testing.md?q=OnTaskRunnerDeleter%20lang:md